sharedbuffers/mapped_struct.py

```diff
@@ -333,14 +335,231 @@ class mapped_dict(dict):
 
     @classmethod
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
-        keys = list(obj)
-        values = [obj[k] for k in keys]
-        return mapped_list.pack_into([keys, values], buf, offs, idmap, implicit_offs)
+        return proxied_dict.pack_into(obj, buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        proxy = proxied_dict.unpack_from(buf, offs, idmap)
+        return proxy.copy()
+
+
+@cython.ccall
+@cython.locals(code=cython.ulong, nbits=cython.ulong)
+def _hash_rotl(code, nbits):
+    return (code << nbits) | (code >> (32 - nbits))
+
+
+@cython.ccall
+@cython.locals(code1=cython.ulong, code2=cython.ulong)
+def _mix_hash(code1, code2):
+    return _hash_rotl(code1, 5) ^ code2
+
+
+_TUPLE_SEED = 4110121799
+_FSET_SEED  = 3377403095
+
+@cython.ccall
+@cython.locals(hval=cython.ulong)
+def _stable_hash(key):
+    if isinstance(key, basestring):
+        hval = xxhash.xxh64(safe_utf8(key)).intdigest()
+    elif isinstance(key, int):
+        hval = key
+    elif isinstance(key, float):
+        trunc_key = int(key)
+        if trunc_key == key:
+            hval = trunc_key
+        else:
+            mant, expo = math.frexp(key)
```

Comment:
Can you make Cython use C's native frexp?

You'll probably have to add an `if cython.compiled` here, because C's frexp is radically different in calling convention. But it should be worth the trouble.

Reply:
I don't know about this. In additon to frexp having an output parameter, which I don't know how it translates to cython, there's the complication that we'd also have to deal with +/Inf and NaN.

Looking at the source, it appears that CPython uses a straightforward translation of C's native call, but at the same time handling the above problems:

https://github.com/python/cpython/blob/2.7/Modules/mathmodule.c#L1155

I can try to cythonize this, but I don't think it's really worth it. Besides, how many floats do we see as keys in hash tables anyway ?